CondFormats/DataRecord/interface/HGCalMappingModuleRcd.h

```diff
@@ -0,0 +1,29 @@
+#ifndef CondFormatsModuleRcd_HGCalMappingModuleRcd_h
+#define CondFormatsModuleRcd_HGCalMappingModuleRcd_h
+// -*- C++ -*-
+//
+// Package:     CondFormats/DataRecords
+// Class  :     HGCalMappingModuleRcd
+//
+/**\class HGCalMappingModuleRcd HGCalMappingModuleRcd.h CondFormats/DataRecords/interface/HGCalMappingModuleRcd.h
+
+ Description: [one line class summary]
+
+ Usage:
+    <usage>
+
+*/
+//
+// Author:      Pedro Vieira De Castro Ferreira Da Silva
+// Created:     Mon, 13 Nov 2023 12:02:11 GMT
+//
+
+#include "FWCore/Framework/interface/DependentRecordImplementation.h"
+#include "FWCore/Utilities/interface/mplVector.h"
+#include "CondFormats/DataRecord/interface/HGCalMappingModuleIndexerRcd.h"
+
+class HGCalMappingModuleRcd
```

Comment:
Do you expect different IOVs from `HGCalMappingModuleIndexerRcd` or `HGCalMappingModuleIndexerRcd` ?

Reply:
No, the IOV will be a full run

Reply:
OK, then you do not need to use different records for these conditions.

You can use a single record for all the conditions that will have the same IOVs.

Reply:
OK so this clears a source of confusion at least for me. I think all this PR can live with a single record for both the indexers and the cell/module characteristics. I'll rename it as HGCalElectronicsMappingRcd and remove the current 3 records. I just tested locally and it works fine.

Reply:
> > You can use a single record for all the conditions that will have the same IOVs.
>
> OK so this clears a source of confusion at least for me. I think all this PR can live with a single record for both the indexers and the cell/module characteristics. I'll rename it as HGCalElectronicsMappingRcd and remove the current 3 records.

While this is true from the framework point of view, there is one caveat though, in relation to the question in https://github.com/cms-sw/cmssw/pull/43751#issuecomment-1952801195 about the inclusion of the payloads to the Global Tag. My understanding is that the CondDB infrastructure supports exactly one data product type per Record. I.e. **if** the data products in these records (that are currently read from files) are eventually intended to be eventually stored in the CondDB, it _might_ be more future proof to use separate Records already now. But of course things can be changed later as well.

Reply:
Ah, my bad, I wasn't aware of this limitation.

I see that in a global tag we can have multiple tags/payloads in the same record, as long as they have different labels:
![image](https://github.com/cms-sw/cmssw/assets/4069793/3f5af292-776d-47ce-9006-d0ba6b9a20f6)

I do not know if they can have different types, though.

Reply:
oh.... ok, then we need 4 records as there are 4 four products.

Reply:
> Is the plan to eventually add the HGCalMappingModuleIndexer and HGCalMappingCellIndexer into CondDB? And after that, would the HGCalMappingIndexESSource not be needed in production setup (but only for tests and uploading new payloads to CondDB)?

Yes, but with a longer time scale for  HGCalMappingModuleIndexer

> Currently the creation of HGCalMappingModuleIndexer depends on HGCalMappingCellIndexer. When(/if) these are moved to CondDB, would the two data products / payloads be completely independent then (except maybe when new CondDB payloads are prepared)?

This flexibility is for now needed because of the large variety of test setups during construction/commissioning. This will crystallize into a final mapping when we install the detector. But we are not yet there

> In the current code the HGCalMappingCellRcd / HGCalMappingModuleRcd could be removed, and HGCalMappingCellIndexerRcd / HGCalMappingModuleIndexerRcd could be used directly instead

You mean overloaded with the corresponding SoA's payloads? I'm also getting confused then. Shouldn't we prepare the records such that it gives us the flexibility we need when we adopt CondDB which would mean 1 format = 1 record?

> Is it foreseen the HGCalMappingCellRcd / HGCalMappingModuleRcd would ever need to depend on any other Record than HGCalMappingCellIndexerRcd / HGCalMappingModuleIndexerRcd?

No, not foreseen

Reply:
> > In the current code the HGCalMappingCellRcd / HGCalMappingModuleRcd could be removed, and HGCalMappingCellIndexerRcd / HGCalMappingModuleIndexerRcd could be used directly instead
> 
> You mean overloaded with the corresponding SoA's payloads? I'm also getting confused then. Shouldn't we prepare the records such that it gives us the flexibility we need when we adopt CondDB which would mean 1 format = 1 record?

So would the plan be to eventually add also `HGCalMappingCellParamHostCollection` and `HGCalMappingModuleParamHostCollection` to CondDB?

Storage of `PortableHostCollection<T>` to CondDB is not supported today, and has not been in the plans. Given the constraints of CondDB (no schema evolution, i.e. once a payload type is added it has to stay like that forever, and the strong link between the payload type and the Record) I would not look into supporting that. Instead, the payload types (to be) stored in CondDB would be prepared with the DB constraints in mind, and converted to GPU-consumable data products in ESProducers. (there are also other examples where a DB data type is converted to a "user friendly" type in similar fashion)

(referring to the discussion in https://github.com/cms-sw/cmssw/pull/43751#discussion_r1497883609) Assuming I understood correctlyt that the data of all the current text files would be eventually be moved to CondDB, it seems like 4 different payloads (and hence 4 Records) would be needed anyway, and for that purpose the present arrangement of the Records seems adequate.

It could be useful to add comments to `HGCalMappingCellRcd.h` and `HGCalMappingModuleRcd.h` along it is foreseen the separate Records will be needed in the future, and with the present code arrangement the dependence on one other Record is necessary.

Reply:
Right i'm aware SoAs can't be ported to CondDB, what is needed in the CondDB is the information needed to fill the SoAs. Ok i'll update the comments on the Rcd files.

Reply:
Thanks. This implies then that when the text files that are currently used by `HGCalMappingCellESProducer` / `HGCalMappingModuleESProducer` are moved to CondDB, the `HGCalMappingCellRcd` / `HGCalMappingModuleRcd` Records need to be made to depend also on the Records of the corresponding future Records (that are used for CondDB). So defining `HGCalMappingCellRcd` / `HGCalMappingModuleRcd` as dependent records already now would be in the right direction.

Reply:
So the framework recommended way to handle the EventSetup would be to have ES modules which read the text file and put the data into the EventSetup and then have the ESProducer read that data to create the needed data product. Then when it is time to switch to using CondDB, CondDB would be a drop in replacement for the ES modules that read the text file.